### ISSUE TITLE
Grading/Requirements for v2

### ISSUE THREAD

**Issue by johngtimms** [usc-csci201-fall2013/simcity201#72]
Quite a few people have been (and still are) confused about the requirements and grading for v1. Since the grading has yet to begin, I think now would be a good time to discuss the exact requirements for v2.

The professors do not desire to provide a rubric to give students experience working with nebulous requirements. However, I think it would be helpful to provide a list (if not exhaustive, then partial) of individualized team-specific requirements for v2 along with the grade reports for v1.

Because very few v1 projects will look the same (due to the flexible requirements) any general requirements published for v2 may not be relevant to all teams. I simply want to have a fair opportunity to do as well as possible. Just like in industry where programmers either get an "A" or an "F" for complete or incomplete work, I know from experience as an intern that companies _also_ hire writers and analysts to ensure that requirements _are_ properly specified and understood by programmers.

@crowley1958 @dwilczyn

**Comment by gabelew** [comment]
He has already mentioned in class that there will be a rubric for v2.

**Comment by dwilczyn** [comment]
I am working on a full and complete rubric for v2. It will be coming out soon.

**Comment by diegovb** [comment]
:D

**Comment by lshiffer** [comment]
^_^

**Comment by eerie444** [comment]
:っ)

**Comment by AnthonyCCH** [comment]
:DDD

**Comment by zachvp** [comment]
= ]

**Comment by ztanner** [comment]
Smiling at 4:40AM is difficult

**Comment by diegovb** [comment]
:c

**Comment by eerie444** [comment]
@ztanner :っ(

**Comment by zachvp** [comment]
=/

**Comment by AnthonyCCH** [comment]
:|

**Comment by mgoulis** [comment]
D=

**Comment by ztanner** [comment]
:baby: :baby_bottle: :cry:

**Comment by kushaank** [comment]
:'(

**Comment by yixincai** [comment]
T_T

**Comment by jeffersonluu** [comment]
ಠ_ಠ

**Comment by AnthonyCCH** [comment]
つ ◕_◕ ༽つ  Give Rubric つ ◕_◕ ༽つ

**Comment by johngtimms** [comment]
![Overjoyed](http://i.imgur.com/wP2Emx1.gif)

**Comment by AnthonyCCH** [comment]
░░░░░░░░░░░████ 
░░░░░░░░░░░█░░█ 
░░░░░░░░░░█░░░█ 
░░░░░░░░░█░░░░█ 
██████▄▄█░░░░░███▄ 
▓▓▓▓▓█░░░░░░░░░░░░█ 
▓▓▓▓▓█░░░░░░░░░░░░░█ 
▓▓▓▓▓█(づ｡◕_◕｡)づ ░░░░░█ 
▓▓▓▓▓█Give Rubric ░░░░░░█ 
▓▓▓▓▓█░░░░░░░░░░░░░█ 
▓▓▓▓▓█░░░░░░░░░░░░█ 
▓▓▓▓▓█░░░░░░░░░░░█ 
▓▓▓▓▓████████████

**Comment by eerie444** [comment]
░░░░░░░░░▄░░░░░░░░░░░░░░▄░░░░ wow so happy
░░░░░░░░▌▒█░░░░░░░░░░░▄▀▒▌░░░
░░░░░░░░▌▒▒█░░░░░░░░▄▀▒▒▒▐░░░ much rubric
░░░░░░░▐▄▀▒▒▀▀▀▀▄▄▄▀▒▒▒▒▒▐░░░
░░░░░▄▄▀▒░▒▒▒▒▒▒▒▒▒█▒▒▄█▒▐░░░
░░░▄▀▒▒▒░░░▒▒▒░░░▒▒▒▀██▀▒▌░░░ 
░░▐▒▒▒▄▄▒▒▒▒░░░▒▒▒▒▒▒▒▀▄▒▒▌░░
░░▌░░▌█▀▒▒▒▒▒▄▀█▄▒▒▒▒▒▒▒█▒▐░░
░▐░░░▒▒▒▒▒▒▒▒▌██▀▒▒░░░▒▒▒▀▄▌░
░▌░▒▄██▄▒▒▒▒▒▒▒▒▒░░░░░░▒▒▒▒▌░
▀▒▀▐▄█▄█▌▄░▀▒▒░░░░░░░░░░▒▒▒▐░ very 5 am
▐▒▒▐▀▐▀▒░▄▄▒▄▒▒▒▒▒▒░▒░▒░▒▒▒▒▌
▐▒▒▒▀▀▄▄▒▒▒▄▒▒▒▒▒▒▒▒░▒░▒░▒▒▐░
░▌▒▒▒▒▒▒▀▀▀▒▒▒▒▒▒░▒░▒░▒░▒▒▒▌░
░▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒░▒░▒░▒▒▄▒▒▐░░
░░▀▄▒▒▒▒▒▒▒▒▒▒▒░▒░▒░▒▄▒▒▒▒▌░░
░░░░▀▄▒▒▒▒▒▒▒▒▒▒▄▄▄▀▒▒▒▒▄▀░░░
░░░░░░▀▄▄▄▄▄▄▀▀▀▒▒▒▒▒▄▄▀░░░░░
░░░░░░░░░▒▒▒▒▒▒▒▒▒▒▀▀░░░░░░░░

**Comment by Rancho06** [comment]
O_OO_OO_O

**Comment by Yunying** [comment]
_^_^_

**Comment by dablosmo** [comment]
(╯°□°）╯︵ ┻━┻

**Comment by xulander** [comment]
(ー_ー)!!　(-.-)　(-_-) ( 一一)　(；一_一)

**Comment by dhata** [comment]
;A;

**Comment by paigekehoe** [comment]
woohoo

¨ ¨ ¨ ¨ ¨ ¨ ¨▲.︵.▲
¨ ¨ ¨ ¨ ¨ ¨ ¨◄ƒ░░ 0 }…..︵.
¨ ¨ ¨ ¨ ¨ ¨◄ƒ░░░░░░ o”)
¨ ¨ ¨ ¨ ¨ ◄ƒ░░░(────.·^”’
¨ ¨ ¨ ¨ ¨ ◄ƒ░░░§ ´
)¨ ¨ ¨ ¨ ◄ƒ░░░░§ ¨ ¨ ¨✺✺✺
) ¨ ¨ ¨◄ƒ░░░░░§¨ ¨ ¨|//✺
);; ¨ ¨ ◄ƒ░.︵.░░░§**(((
);;;¨ ¨◄ƒ░(░░)____**//
);;;;¨◄ƒ░░░░░░__//
);;;;;◄ƒ░░░░░░░░§
);;;;;;;░░░░░░░░░(((

**Comment by dwilczyn** [comment]
Thank you for the enthusiasm. Here is the rubric. I will leave it open for comment for a few days after which it will be put on the website. [Sorry for the way the formatting came out below. It's Git's fault.]

Team Rubric for SimCity201 v2
Each NUMBERED item below represents 2½ points of grade; partial credit is available. Your team starts out with 100 points and you lose 2½ points for each item which doesn’t comply.  The idea is that:
• You can still get an A- team grade with four flaws.
• You can still get a B- team grade with eight flaws.
• etc.
Normative Scenarios – Baseline, i.e., little or no interleaving  
Scenario: [Tests all the behaviors.]
1.  All workplaces (markets, all restaurants, banks) fully employed.  
2.  Day starts and all workers go to work. 
3.  One not-working person eats at home, then visits all the workplaces by walking.
4.  Roads should have appropriate complexity [e.g. intersections with stop signs and/or signals]
Scenario: [Tests all the behaviors.]
5.  All workplaces fully employed.  
6.  Day starts and all workers go to work. 
7.  Three not-working persons eat at home, then visit all the workplaces in different orders. [one should walk, one should take a car, one a bus.]
Scenario: [Tests cook, cashier, market interaction.]
8.  Each restaurant gets low on food and orders from market(s).  
9.  Market delivers food to the open restaurant.
10. Market sends invoice to cashier, who verifies and pays it.
Scenario: For large teams [Tests party behavior.]
11. Show a party being planned.
12. All invitees RSVP before RSVP-date.
13. At least 2 friends should RSVP “yes.”
14. Throw the party. 
Scenario: [Shows bus-stop behavior]
15. Person visits a bus stop. [as part of 7]
16. Bus arrives.
17. Person gets on.
18. Person gets off at destination.
 
Non-Normative Scenarios – Baseline, i.e., minimum interleaving  
Scenario: [Shows that people know they can’t visit certain workplaces.]
19. Bring each workplace down, one by one. 
20. Show how one not-working person still visits all the workplaces but not the ones that are down.  Say you only have one bank and it is down, the person should avoid all banking behavior. 
Scenario: [Tests market behavior]
21. Market delivery fails because restaurant is closed. 
22. Market redelivers when restaurant is open.
23. Market sends invoice to cashier, who verifies and pays it.
Scenario: For large teams [Tests party behavior.]
24. Show a party being planned.
25. Some invitees do not RSVP before RSVP-date.
26. These flakes should be reminded.
27. At least 2 friends should RSVP “yes.”
28. Throw the party. 
Scenario: For large teams [Tests party behavior.]
29. Show a party being planned.
30. Party is canceled because no one is coming.
Normative Scenarios – Normal operation, i.e., normal interleaving
Scenario:
31. Fully employed workplaces [enough people to fill all the workplace roles].
32. Fully populated city [enough people to be wandering around doing interesting things]. At least  50 people.
33. All restaurants integrated. 
34. At least 2 instances of other workspaces (your decision as to how many more you want).
35. People make decisions to eat at home.
36. People make decisions to eat at particular restaurants.
37. Enough vehicle traffic to show stopping/starting at intersections.
38. Vehicles stop for pedestrians with right of way.
Animation Details
39. Icons/images have identity information to help track behavior during grading.
40. Icons/images are beautiful for non-grading view, but you do not have to do the artwork.
Gui Controls
41. Screen to run full scenarios, i.e., pick a scenario to run.
42. Screens to add people to the simulation and give them properties.
43. Screens to add properties to workplaces, i.e., set inventories, etc.
44. Incorporation of log filter mechanism described by CP Keith DeRuiter.
Runtime Behavior
45. Animation should not freeze.
46. No error exceptions. [If in a restaurant, then it is an individual deduction.]
47. Routing behavior must show that agents don’t bump into buildings, tables, etc.
48. All moving objects must move on appropriate venues (vehicles parked in garages/roads, move on roads, people on sidewalks and crosswalks).
49. Vehicles should behave appropriately at intersections.
Design Docs
50. Quality of normative interaction diagrams.
51. Quality of non-normative interaction diagrams.
Non-Normative Scenarios – 3 points a scenario; you design them
Scenario: Bank Robbery
52. An evil person decides to rob a bank.
53. On entrance he … (you design it)
Scenario:  Vehicle accident
54. Two vehicles collide.
55. Then… (you design it).
Scenario: Vehicle hits pedestrian
56. A vehicle hits a pedestrian.
57. Then… (you design it).
Scenario:  Weekend behavior is different
58. Some workplaces are closed on the weekend.
59. Show that people naturally avoid them and seek alternatives.
Scenario:  Job Changing/Firing/Shifts
60. Each workplace shows how its employees can change during working hours.
61. The change can be the result of a shift change, a firing, or quitting/hiring.
Scenario: Extra Credit
62. A person gets very, very hungry.
63. Unfortunately, there is no food in his house and he has no money.
64. He goes to each market, looking for food in the dumpster. He finds none.
65. He goes to each restaurant, knowing they will feed him and he can flake out on the bill. 
66. Unfortunately, each host remembers him and refuses him entry.
67. Weak from hunger, he collapses on the sidewalk.
68. A small group of people gather trying to help him.
69. Unfortunately, he dies anyway.
70. The group is angry that such a thing could happen in their city and form a protest.
71. The police come to break up the protest while an ambulance arrives to remove the body.
72. The body is taken to the morgue where the coroner decides the brain is still in good enough condition for a transplant.
73. A professor named Wilczynski was told by his colleagues that he needed to have his head examined for giving team projects to sophomores.  
74. At that examination the doctor said nothing could be done for his condition, so Wilczynski applied for a brain transplant.
75. When the coroner called to tell him a brain is available, Wilczynski rethinks the problem and decides that writing a rubric for v2 is easier than having the surgery. However, he tells the coroner to keep the brain on ice if he changes his mind in case the rubric is received badly by his students.
76. The students read the rubric and ... [you finish the scenario].
Rubric for Individual Grades
Here EACH of your components is subject to the following: 
77. Quality of your pseudo-code for the design of your component.
78. Quality of agent/role code [Does your code conform to agent methodology? Does it work? Can graders understand it?]
79. Quality of gui code (assuming you did some).
80. Quality of Unit Testing [tests obviously depend on complexity of the component].
And finally:
81. Division of work. Did you do your fair share of work?
This last point is a tough one. Obviously, the PersonAgent represents the biggest piece of work at the agent level. All parts of the gui are critical and difficult. If all you did, for example, is write the Bank roles, your overall grade will substantially lower than those on your team who did more. I suspect there will be more to say about this topic.

**Comment by dwilczyn** [comment]
I forgot the rubric for the producer-consumer waiter. Here it is:

Proper integration of producer-consumer code
82. Having WaiterAgent base class extended by two subclasses for normal waiter and producer-consumer waiter.
83. Having a non-random way to create waiters of each class, e.g. from gui, from configuration file.
84. Have a scenario to show normal waiter operates successfully, e.g. unit test, or ...
85. Have a scenario to show producer-consumer waiter operates successfully, e.g. unit test, or ...

**Comment by eerie444** [comment]
Question about producer-consumer waiters: are all restaurants expected to implement this new design, or is it acceptable for some restaurants to do so while some retain the normal waiter functionality? It doesn't seem to reflect reality that within a single restaurant, some waiters talk directly to the cook while some clip their orders to a rotating shared data location.

**Comment by dwilczyn** [comment]
This is an individual assignment and has nothing to do with reality. I want each student to do the refactoring required to make this behavior happen.

**Comment by johngtimms** [comment]
Is it intentional that numbered items \* 2.5 does not equal 100?

"Market delivery fails because restaurant is closed." 
We designed our city to be open 24 hours and nothing is ever closed. Do we need to redesign to have regular open/closed cycles daily (i.e. beyond some buildings being closed on weekends) or (since this is a non-norm) can we simply set the restaurant closed at some point in the scenario?

"People make decisions to eat at particular restaurants."
Does this mean that the choice of restaurant should not be random? Is "I'm closest to Restaurant X so I'll eat there." sufficiently non-random?

"Icons/images are beautiful for non-grading view."
I don't want our team to be distracted from coding by finding or developing graphics for this requirement. We're learning to be programmers, not graphic designers. The intent here is obviously to show that teams have the ability to work with GUI elements beyond colored squares. Can this requirement be changed or can the professors provide a set of images suitable for use to the whole class?

**Comment by dwilczyn** [comment]
1) Yes.
2) Yes, there should be opening and closing hours. But, I doubt you have to redesign anything.
3) Random is ok for v1, not v2. Closest is one of many possible decisions, but will produce only very predictable behavior.
4) Maybe "beautiful" was too strong a word. Maybe post this as a separate issue and see what the class comes up with.

**Comment by Valakor** [comment]
@johngtimms 
You don't need to work too hard to find suitable art, I don't think. A quick search of Google images for "sprite sheet" will get you tons of 2D artwork that you can use. Take 10 minutes to crop out pictures using Paint or something and you should be fine. I don't think the professors wanted us to produce our own artwork for this project?

**Comment by johngtimms** [comment]
@Valakor I already saw some people doing that for v1. Hopefully it can be made clear in the rubric that producing your own artwork isn't necessary.

**Comment by dwilczyn** [comment]
You do not have to produce your own artwork. I'll add that to the rubric.

**Comment by Cherys** [comment]
Would a smaller team also need at least 50 people in their city?
By "At least 2 instances of other workspaces", do you mean that instead of having one market, we could have 3 and that would satisfy this requirement?
Does all functionality from previous restaurant projects need to still be in place, or can we edit it to make it work better with our cities? For example, certain things like setting a customer to "hungry" are no longer possible (right?) Will we get penalized for missing requirements from previous restaurant versions even if the restaurant fills its function adequately?

**Comment by Cherys** [comment]
*When I asked about smaller groups perhaps not needing to implement quite so many people, that came from the fact that we will have significantly fewer jobs for them. A group of four versus a group of seven means that the former group will have three fewer restaurants than the latter, which makes up a large percentage of overall jobs

**Comment by shanemileham** [comment]
@dwilczyn - Regarding rubric point number 4 (Roads should have appropriate complexity [e.g. intersections with stop signs and/or signals]), this seems to contradictory to our entire GUI that our team designed. As John said in the first post about the flexibility of V1 requirements leading to different products, we designed our city in a different way which allowed us a more efficient movement algorithm and city layout. Having to incorporate intersections would be fine if it were a requirement from the beginning, but having to redesign our entire city, building layout, transportation mechanism, and movement algorithm at this late stage seems more difficult than intended for the assignment.

Would there be a way to substitute a more personalized road requirement (for example, we could include sidewalks, alleys, and bus lanes) which could teach us more than redoing all of our existing GUI work?

Thanks for your understanding.

**Comment by dwilczyn** [comment]
- 50 people in the city is not a lot, smaller team or otherwise. Plenty of waiting rooms in the workspaces.
- 2 instances of each workspace.
- Setting customer to hungry is simply replaced by setting a person to various states of hunger. Waiters still need to go on break. All the other restaurant non-norms should work.
- Prof. Crowley and I have drawn many, many pictures of intersections throughout the early stages of projects, talking about pedestrians moving though crosswalks. Not having intersections is a flaw.

**Comment by dwilczyn** [comment]
The rubric will be published on the website at the end of the day today. Last chance for comments.

**Comment by alryco** [comment]
Will the performance/framerate of our city be taken into account in grading? Our team has not been too concerned with optimization for v1. Right now with 10-15 people and only 1 restaurant, our city has a relatively low framerate and I am concerned with how well our city will be able to run with 50+ people running simultaneously. Is this something that will have an affect on grading?

**Comment by richy0101** [comment]
@dwilczyn I still don't understand why there are 85 requirements for v2 that are worth 2.5 points each but total points is 100? 
212.5 != 100?

**Comment by ztanner** [comment]
I haven't done the math but there are extra credit scenarios in that list of 85.

**Comment by alghanmi** [comment]
@richy0101, please read what @dwilczyn wrote a bit more carefully:

> Each NUMBERED item below represents 2½ points of grade; partial credit is available. Your team **starts out with 100 points** and **you lose 2½ points for each item which doesn’t comply**.

In other words, we will not grade by adding-up the score, but with deducting points when things go wrong.

**Comment by dwilczyn** [comment]
The extra-credit scenario was, obviously, a joke. It won't be in the version soon to be posted.

**Comment by jdgreenberger** [comment]
The problem with this grading scheme is that you could hypothetically get less than 0 points, and have to do more than 100 points worth of work in order to get 100. Why don't you just weight the requirements to add up to 100, or start with a higher score?

**Comment by ecoutre** [comment]
I agree that it doesn’t make sense that you can lose more points than you can earn.

**Comment by crowley1958** [comment]
Frame rate will not be taken, as long as the grader can tell that things are happening properly.

**Comment by pbradshawusc** [comment]
(34) For the bank, should we implement 2 or more instances? When discussing the design prior to v1 and during v1, my team was under the impression that having only one bank was sufficient for v2 requirements. (I can't think of the issue or comment in class specifically at the moment as to why we were under this impression, but I believe there is an issue of it on GitHub and can search for it if required). Due to this, we built our design around only having one bank (as a part of the person/bank interaction and deciding what bank to go to). Is this still the case or should we redesign our banks/people to implement multiple banks as well as multiple bank accounts?

(44) For implementing the log filter mechanism, we already have separate consoles for each building, visible when the building is open. This allows a feed much like for the restaurant. Is this acceptable instead of the log filter? It seems to accomplish the same purpose of cleaning up the log and it would seem redundant to add the log filter inside of every building as there are few messages to filter through.

(47) For agents bumping into tables, it was never posted as a requirement or guideline to have A\* or any sort of collision avoidance algorithm within our personal restaurants. Does this requirement apply to our individual restaurants as well? I fear this addition to the individual portion of the grade will take team members further out of the team portion of the project (as A\* would have to be tuned for each member's GUI) and would hurt the team as a whole. 

(82) Also, as a personal question for my waiters, my producer-consumer waiter and regular waiter are very similar and I was able to accomplish the two with a single class and an enum for the type rather than creating two new classes. In addition, my toggle method meets requirements 83-85. I can see how larger variations between the waiters would require the extra classes, but with just a few lines of code difference as the end product, would this solution also be acceptable/efficient?

**Comment by alghanmi** [comment]
@jdg123, @ecoutre and to all @usc-csci201-fall2013/students,

In [my discussion about grades and grading](https://github.com/usc-csci201-fall2013/simcity201/issues/71#issuecomment-29273829) I expressed in a well mannered way what to expect from us as a teaching staff and the best way - based on experience - you, as students, should behave when it comes to grades. If you did not read that, please go back and give it a look in https://github.com/usc-csci201-fall2013/simcity201/issues/71#issuecomment-29273829.

Now, if you are not into stuff like that and you just really care about grades and how to get as many points as you can regardless of the process or the educational experience and irrespective of what we tell you, here is my comment to you:

**YOU** asked for a rubric and against the better judgment of most of the teaching staff, **you got one**.

After getting the rubric, you are complaining that it is not **your rubric**, it is not **tailored to your case** or your **team will suffer** if we followed this one. Well, _**this is what we have been telling you since you started asking for a rubric**_. You can't have it both ways.

You want us to trust that you did a 100% work and deserve a 100% credit or close to it. We - and by we I mean the teaching staff - also know how to grade and we asked you to trust us. The department trusted us to do this right, the school trusted us to do this right, the university trusts us to do this right. We believe we can a very good and more importantly fair job grading you because **we have the track record to prove it**. Years of teaching experience between the professors, TAs and CPs. Sorry, DECADES. We know the course material (we taught it before), we know large student projects (we did them before), we know how to grade (we do it all the time).

After all this, if you do not trust us with your grades, it is your prerogative and I respectfully tell you: You are wrong and show some respect. It may be years down the line, but you WILL look back at this stage of your college career and say "I wish I did this differently."

Professor @dwilczyn published the rubric and made it available for open comment. We got your comments and are incorporating them in the final version of the rubric. You are also welcome to commend and discuss the expectations and merits of each point.

In conclusion, I want to remind you that developers are people who are skilled with multiple aspects of the software development process. They are skilled (to different degrees of proficiency) at research, requirements collection, design and testing. A programmer who writes code to as a chore or a repetitive process without much problem solving is called [something else](http://en.wikipedia.org/wiki/Code_monkey) and it is [not developer](https://www.youtube.com/watch?v=OcdjxpN3FTw) and definitely not engineer.

**Comment by waynchi** [comment]
If we want to run a scenario, it will be fine if it restarts the simulation and starts running that certain scenario right.

**Comment by waynchi** [comment]
Also, if we actually do the Extra Credit portion, can we get extra credit?

**Comment by chrismclaughlin55** [comment]
Was the official rubric ever posted? If so, where was it posted? Or is it just what was posted earlier in this issue thread?

**Comment by johnsonh** [comment]
For point 75, does the ice need to melt based on a timer or can we assume that the ice remains indefinitely cold?

**Comment by luca992** [comment]
Honestly I have never complained once about any of these grading issues. I
am not paying to attend this school to have my instructors accuse me of
such things. I don't appreciate basically being yelled at and possibly
punished grade wise for other students actions. I don't care only about
grades, and I am here for the education. But I'm sorry, get a grip... this
course clearly needs more organization or there wouldn't be such chaos.

**Comment by BachDinh1994** [comment]
Luca, I also did not complain about the grading issues and I did not ask for a rubric either. Therefore, I don't appreciate getting accused. However, Rami's right about the fact that we need to show some respect. If we're struggling, then it's pretty much our fault due to poor time management and/or lack of communication between team members. Professors and TAs want you to succeed, and because we're all qualified by having made it this far in 201, there are no excuses as to missing deadlines and/or complaining about requirements. It's true that this project was much more time-consuming than Prof. Wilczynski had originally anticipated, but it was also a lot more interesting. He wouldn't have made such a large rubric set in stone had we shown some respect and complied instead of constantly complaining. Having said that, I do believe the deadline next week is a little early, considering it's only a week after an enjoyable thanksgiving. I'd be infinitely obliged to the TAs, CPs, Prof. Crowley and Prof. Wilczynski if we could push it a little further down during that week. A couple of days will make all the difference.

**Comment by lee533** [comment]
@johnsonh That made my day, thank you a ton.

**Comment by AnthonyCCH** [comment]
@johnsonh definitely the best CSCI201 joke lmaoooo

**Comment by dwilczyn** [comment]
Let's all get a grip here. This is in effect a new course. Never have csci201 students come in to this class with so little experience. The csci200 buffer class contributed a lot to the programming experience expected of csci201 students. I'm not going to make a big deal about this, but this issue is the "elephant in the room."

Let me try to answer each of the posts.
- A few wonder about the grading system and whether is makes sense to be able to lose more points than you can earn. The answer is: Yes, it makes sense.
- My extra credit was an attempt at humor. Go ahead and try if you have the time. Most teams in the past had fun with the project. We had many successful attempts at funny non-norms. 
- Here are answers to a series of questions you have asked:
  (34) For the bank, should we implement 2 or more instances? When discussing the design prior to v1 and during v1, my team was under the impression that having only one bank was sufficient for v2 requirements. (I can't think of the issue or comment in class specifically at the moment as to why we were under this impression, but I believe there is an issue of it on GitHub and can search for it if required). Due to this, we built our design around only having one bank (as a part of the person/bank interaction and deciding what bank to go to). Is this still the case or should we redesign our banks/people to implement multiple banks as well as multiple bank accounts?

Ans: Many of you wrote code in the restaurant that made INCORRECT assumptions about having ONE instance of an agent, e.g., your v2.2 market assuming only 1 restaurant as a customer was an example. This requirement attempts to make sure this erroneous assumption is not part of your SimCity201. 

(44) For implementing the log filter mechanism, we already have separate consoles for each building, visible when the building is open. This allows a feed much like for the restaurant. Is this acceptable instead of the log filter? It seems to accomplish the same purpose of cleaning up the log and it would seem redundant to add the log filter inside of every building as there are few messages to filter through.

Answer: If you have your own mechanism, just document that it is an alternative to that stated in the rubric. In general that is the case for every rubric statement. If you want clarification, just ask. Without the rubric all such alternatives would have been gladly accepted. But the rubric memo-izes everything and makes everyone crazy with fear (as well as making everything too uniform). 

(47) For agents bumping into tables, it was never posted as a requirement or guideline to have A\* or any sort of collision avoidance algorithm within our personal restaurants. Does this requirement apply to our individual restaurants as well? I fear this addition to the individual portion of the grade will take team members further out of the team portion of the project (as A\* would have to be tuned for each member's GUI) and would hurt the team as a whole.

Answer: Don't have to use A*, but an animation where things are bumping into things is clearly inferior to one that doesn't. Don't worry about it if you don't want to. It's just a 2 1/2 point deduction. The question you should be asking is whether it is "A" work or "B" work or ... Rami has tried to instill in you a sense of YOU MAKING your own decision about what is A work, not complying with an obviously impossible-to-write rubric.

(82) Also, as a personal question for my waiters, my producer-consumer waiter and regular waiter are very similar and I was able to accomplish the two with a single class and an enum for the type rather than creating two new classes. In addition, my toggle method meets requirements 83-85. I can see how larger variations between the waiters would require the extra classes, but with just a few lines of code difference as the end product, would this solution also be acceptable/efficient?

Answer: Of course it is trivial to use a boolean or enum to decide what waiter to use. That is NOT the point. I want all of you to flex your programming muscles and do the refactoring and build abstract base classes--that is the point.

As for point (75) ice melting obviously needs modeling for the scenario to work. You can only dither so long while the brain is on ice. A timer is fine.

As to another extension...it is too late for that, really.

**Comment by Cherys** [comment]
I whole heartedly agree withthe concept of flexibility allowing for students to have fun with a project and for the exceptional teams to have room to shine, but unfortunately I doubt that any of the teams this year feel they have room to have fun. This is very disappointing, as there are definitely a lot of fun alternate scenarios we could have programmed into a simulation like this if we had the time and inclination. However, I think the operating procedure of virtually all of the current teams is slam out as much work as possible as fast and dirty as possibly to meet the most requirements they can manage. Fun happens when you have leaway, and right now I don't think that any team is sitting comfotable with the assumption that they can complete this project to the full requirement level and then have time left over to play with it. Which is honestly a true shame--this project could have really been fun under different circumstances

**Comment by waynchi** [comment]
I agree with the above. We were going to add a lot of features, but with us
lacking in CS200 Gui experience, we have to dedicate so much of our time to
Gui that we can't make all parts of our project super fun and cool. Perhaps
CS200 will be brought back in the future to help us?

**Comment by khupka** [comment]
With all due respect, Professor @dwilczyn, I have read and understood your points. However, I do not think it is fair to make personal generalizations about students.

1) I am a junior and there are other students in the class who are upperclassmen.
2) I have worked extensively on group projects of different types both in school and in the workplace.
3) Everyone's financial situations are different and are personal matters.

Obviously everyone is upset and I feel like there is a lot of finger pointing. At this point I think that there should be a compromise:

Yes, this course needs more organization
Yes, students need to stop posting immature/sarcastic comments; how does this progress the situation?

Please understand that although students may not express themselves in the best of light, there is a lot of frustration with this project.  I agree, this project has multiple benefits and a lot of lessons to be taken away from it.  I appreciate the effort in trying to make this the best experience for us.  As someone who is recruiting for internships, this is an amazing experience to talk about.  However, seeing as we are less prepared than previous students taking this class, the scope of this project is a lot to handle and detracts from the experience than what was hoped for.

Although we are asked to place experience over grades, it is difficult to do so when even our work in other courses suffer because of the vast amount of work in this class.  I also understand this is trying to simulate the workplace; however, with a full engineering course load, this is overwhelming.  It saddens me to hear that a professor does not care about respect from students. Regardless of any job and differences people may have with each other, people should have mutual respect for each other.  It is my hope that from now until the end of the semester we all act with this in mind.

**Comment by diegovb** [comment]
I feel like I have never agreed this much with anything before.

**Comment by ancb2000** [comment]
I really don't understand why they took csci200 out of the curriculum.

**Comment by ztanner** [comment]
I imagine it's somewhat similar to the idea that if you can't swim and are thrown into the deep end of a pool, you'll probably learn to swim.

**Comment by diegovb** [comment]
Well, it's just my opinion, but I feel like there's more useful things to do than learn Swing to perfection.

**Comment by smoran02** [comment]
I agree with @khupka. While this class might be intended for sophomores, the redesigning of the curriculum has led to several seniors like me being in this class.

**Comment by lshiffer** [comment]
I'm concerned with some of the items in the proposed rubric that are of this nature:
1. One not-working person eats at home, then visits all the workplaces by walking.

Our Person class randomly decides what mode of transportation to take with higher probabilities given to driving if they have a car and walking if the location is considered close.

Moreover, our base Person class does not visit a restaurant if not hungry and only visits the bank/market if there is a need (similar to only visiting a restaurant if hungry or going to work). 

For such scenarios as the one listed above, it would require redesigning our entire Person class with the Person having less 'freedom' which seems to defeat the entire purpose of a simulation -- And doesn't make sense (If Person is not working and if Person has not visited every public location in the city today, then Person can go walk to every place in city, check out respective role, stand inside for [How long?], return the role, walk to next public location, repeat).  Depending on a city's time scale, this could take a Person all day and if such a rule is added to the scheduler can really slow down the program and create a Person mob (depending on the number of Person's added to the city). 

The only other option I see is to have a second Person class designed just to ensure such scenarios occur (which defeats the entire purpose of verifying the actual Person class behavior) or having a flag of some sort in Person that prevents the Person making their own decisions and forces such scenarios to occur (which will litter our Person class and make the code less readable). 

Neither option I'm much of a fan of as it creates a secondary project that doesn't serve any purpose in regards to the actual simulation, and given the time constraints I feel our time can be best spent else where to deliver a polished(!), fully simulated city that we can be proud to add to our portfolio and reflect back on. 

Realistically, I imagine such a scenario being carried out over time by a single Person (Maybe they visit restaurant A, then go get loan, then return home...  Later that Person cooks at home but first goes to the market...  Later that Person goes to a different restaurant when hungry again, etc..) or all such events occurring simultaneously by multiple Persons (PersonA goes to the market while PersonB gets a loan at the bank while PersonC goes to RestaurantD for lunch, and so forth).

**Comment by dwilczyn** [comment]
I certainly don't want you to redesign your person agent for this one scenario. During normal running of a simulation it would never carry out such a scenario. This scenario is more like a unit test, pushing the PersonAgent to visit ALL the sites. It's not a test of the person agent, per se, but a test of all the sites (and perhaps) all the transportation methods.

Can't you carry it out from your gui? Or perhaps a hack forcing it through the scenario. That would be ok.

**Comment by aleenabyrne** [comment]
@kkthxby3 Maybe make another constructor that sets the values you need to force them to eat, go to bank, etc. and override the 'random' functions that set these values? Then you can just call this specific constructor with different values for your scenarios, so it would be reusable and clean.

**Comment by pixelprizm** [comment]
@kkthxby3 Perhaps you could build into your PersonAgent a list of commands (could be just a list of String) and you could add to the top of the role-choosing section of the PersonAgent scheduler a section like this:

```
if(current role finished and commands is not empty) {
   set current role to the role indicated by commands.get(0);
   commands.remove(0);
   set current role to active; //however your code sets roles to active/current
   return true; //also depends on the rest of your scheduler, of course
}
```

Then, to run the PersonAgent without changing its normal behavior, you would just refrain from adding any commands to the list of commands at the beginning.

**Comment by waynchi** [comment]
Hi, my question was overlooked in the mass amount of replies, so here it is again.

Are we allowed to restart the simulation for the scenarios? It would be very strange to have to do the scenarios while your normal city is running.

**Comment by dwilczyn** [comment]
Yes. But, use your judgment. Most scenarios should NOT need a restart of the system.

**Comment by Yunying** [comment]
I have a question about Scenario S. It say a change can be a result of quitting/hiring/firing. Do we have to get all three to work or just choose one of the three? Does it mean that there has to be a scene where one person steps in and one person steps out, in addition to what we did in restaurant like adding waiters?

**Comment by dwilczyn** [comment]
I'm looking for one case where one person agent takes over the role of a different person agent while the simulation is running, One firing/hiring or shift-change should do it.  Just hiring, like adding waiters isn't enough. It's the takeover of an existing role that I want to see.
